File: packages/shared/src/money.ts
/**
 * Money in SnapTab is always an integer count of MINOR units (paise for INR,
 * cents for USD). Floating point never touches an amount: `0.1 + 0.2 !== 0.3`
 * is not a rounding curiosity when four people are arguing over a restaurant
 * bill, it is a support ticket.
 *
 * The wire format, the Postgres columns and the Kotlin models all use the same
 * integer. Formatting to "₹2,575.00" happens at the very edge, for display only.
 */

/** Currencies whose minor unit is not 1/100 of the major unit. */
const EXPONENTS: Readonly<Record<string, number>> = {
  BHD: 3,
  IQD: 3,
  JOD: 3,
  KWD: 3,
  OMR: 3,
  TND: 3,
  CLP: 0,
  ISK: 0,
  JPY: 0,
  KRW: 0,
  VND: 0
};

export function minorUnitExponent(currency: string): number {
  return EXPONENTS[currency.toUpperCase()] ?? 2;
}

export function minorUnitsPerMajor(currency: string): number {
  return 10 ** minorUnitExponent(currency);
}

/** `toMinor('2575.00', 'INR') === 257500`. Accepts numbers or decimal strings. */
export function toMinor(amount: string | number, currency = 'INR'): number {
  const exponent = minorUnitExponent(currency);
  const text = typeof amount === 'number' ? amount.toFixed(exponent) : amount.trim();

  const match = /^(-)?(\d+)(?:\.(\d+))?$/.exec(stripGroupSeparators(text));
  if (!match) throw new RangeError(`not a decimal amount: ${JSON.stringify(amount)}`);

  const [, sign, whole, rawFraction = ''] = match;
  const fraction = rawFraction.padEnd(exponent, '0');
  if (fraction.length > exponent) {
    // More precision than the currency has: round half-up on the first extra digit.
    const keep = Number(fraction.slice(0, exponent));
    const next = Number(fraction[exponent]);
    const rounded = keep + (next >= 5 ? 1 : 0);
    const value = Number(whole) * 10 ** exponent + rounded;
    return sign ? -value : value;
  }
  const value = Number(whole) * 10 ** exponent + Number(fraction || '0');
  return sign ? -value : value;
}

/** `fromMinor(257500, 'INR') === '2575.00'` — a plain decimal string, no symbol. */
export function fromMinor(minor: number, currency = 'INR'): string {
  assertInteger(minor);
  const exponent = minorUnitExponent(currency);
  if (exponent === 0) return String(minor);
  const negative = minor < 0;
  const digits = String(Math.abs(minor)).padStart(exponent + 1, '0');
  const whole = digits.slice(0, -exponent);
  const fraction = digits.slice(-exponent);
  return `${negative ? '-' : ''}${whole}.${fraction}`;
}

export function formatMinor(minor: number, currency = 'INR', locale = 'en-IN'): string {
  assertInteger(minor);
  return new Intl.NumberFormat(locale, {
    style: 'currency',
    currency,
    minimumFractionDigits: minorUnitExponent(currency)
  }).format(minor / minorUnitsPerMajor(currency));
}

export function sumMinor(values: readonly number[]): number {
  let total = 0;
  for (const value of values) {
    assertInteger(value);
    total += value;
  }
  return total;
}

/**
 * Splits `totalMinor` in the given weight ratio with no lost or invented minor
 * units, using the largest-remainder (Hamilton) method: floor every share, then
 * hand the leftover units out one at a time, largest fractional part first.
 *
 * `preferIndex` wins ties, which is how the payer absorbs the odd paise.
 */
export function apportion(
  totalMinor: number,
  weights: readonly number[],
  preferIndex = 0
): number[] {
  assertInteger(totalMinor);
  if (weights.length === 0) {
    if (totalMinor !== 0) throw new RangeError('cannot apportion a non-zero total across nobody');
    return [];
  }
  if (weights.some((w) => w < 0)) throw new RangeError('weights cannot be negative');

  const weightTotal = weights.reduce((a, b) => a + b, 0);
  if (weightTotal === 0) {
    // Nobody carries any weight: fall back to an even split.
    return apportion(totalMinor, weights.map(() => 1), preferIndex);
  }

  const sign = totalMinor < 0 ? -1 : 1;
  const magnitude = Math.abs(totalMinor);

  const exact = weights.map((w) => (magnitude * w) / weightTotal);
  const shares = exact.map((value) => Math.floor(value));
  let leftover = magnitude - shares.reduce((a, b) => a + b, 0);

  const order = exact
    .map((value, index) => ({ index, remainder: value - Math.floor(value) }))
    .sort((a, b) => {
      if (b.remainder !== a.remainder) return b.remainder - a.remainder;
      if (a.index === preferIndex) return -1;
      if (b.index === preferIndex) return 1;
      return a.index - b.index;
    });

  for (let i = 0; leftover > 0; i = (i + 1) % order.length) {
    const slot = order[i]!;
    if (weights[slot.index]! === 0) {
      // Never push a unit onto someone with no weight unless everyone has none.
      if (order.every((o) => weights[o.index] === 0)) {
        shares[slot.index]! += 1;
        leftover -= 1;
      }
      continue;
    }
    shares[slot.index]! += 1;
    leftover -= 1;
  }

  return sign === -1 ? shares.map((s) => -s) : shares;
}

/**
 * Amounts reach us written every which way: `2575.00`, `1,300.00`,
 * `2,57,500.00` (Indian lakh grouping), `1.300,00` (European). Work out which
 * separator is the decimal point, drop the rest, and hand back a plain
 * `1234.56`.
 */
function stripGroupSeparators(input: string): string {
  const text = input.replace(/[\s\u00A0_]/g, '');
  const lastDot = text.lastIndexOf('.');
  const lastComma = text.lastIndexOf(',');

  if (lastDot >= 0 && lastComma >= 0) {
    // Whichever comes last is the decimal separator.
    return lastDot > lastComma
      ? text.replace(/,/g, '')
      : text.replace(/\./g, '').replace(',', '.');
  }
  if (lastComma >= 0) {
    // A lone comma with exactly two digits after it and none of the grouping
    // shape before it is a European decimal comma: `12,50`.
    const isDecimalComma = /^-?\d{1,3},\d{2}$/.test(text) && !/^-?\d{1,3},\d{3}$/.test(text);
    return isDecimalComma ? text.replace(',', '.') : text.replace(/,/g, '');
  }
  if ((text.match(/\./g) ?? []).length > 1) {
    // `1.300.000` — every dot is a group separator.
    return text.replace(/\./g, '');
  }
  return text;
}

function assertInteger(value: number): void {
  if (!Number.isInteger(value)) {
    throw new RangeError(`money must be an integer number of minor units, got ${value}`);
  }
}

File: apps/android/app/src/main/java/com/snaptab/app/core/Money.kt
package com.snaptab.app.core

import java.text.NumberFormat
import java.util.Currency
import java.util.Locale
import kotlin.math.abs

/**
 * Money on this side of the wire is the same integer count of minor units the API
 * and Postgres use — paise for INR. Nothing is ever held as a Double.
 *
 * The old app formatted with `DecimalFormat("#,##0.00")` and a hardcoded "₹" in six
 * different places, which put the symbol on the wrong side for half the world's
 * locales and grouped digits the Western way even in India.
 */
object Money {

    private const val DEFAULT_CURRENCY = "INR"

    private fun exponentOf(currency: String): Int = when (currency.uppercase(Locale.ROOT)) {
        "BHD", "IQD", "JOD", "KWD", "OMR", "TND" -> 3
        "CLP", "ISK", "JPY", "KRW", "VND" -> 0
        else -> 2
    }

    /** `format(257500)` → "₹2,575.00" in en-IN, with correct lakh grouping. */
    fun format(
        minor: Long,
        currency: String = DEFAULT_CURRENCY,
        locale: Locale = Locale.getDefault(),
        showDecimals: Boolean = true
    ): String {
        val exponent = exponentOf(currency)
        val formatter = NumberFormat.getCurrencyInstance(locale).apply {
            runCatching { this.currency = Currency.getInstance(currency.uppercase(Locale.ROOT)) }
            minimumFractionDigits = if (showDecimals) exponent else 0
            maximumFractionDigits = if (showDecimals) exponent else 0
        }
        val divisor = Math.pow(10.0, exponent.toDouble())
        return formatter.format(minor / divisor)
    }

    /**
     * Drops the paise when they are zero: "₹2,575" rather than "₹2,575.00". Used in
     * lists and on notifications, where the decimals are noise.
     */
    fun formatCompact(minor: Long, currency: String = DEFAULT_CURRENCY, locale: Locale = Locale.getDefault()): String {
        val exponent = exponentOf(currency)
        val unit = Math.pow(10.0, exponent.toDouble()).toLong()
        return format(minor, currency, locale, showDecimals = minor % unit != 0L)
    }

    /** Signed, for a ledger row: "+₹643" or "−₹486" with a real minus sign. */
    fun formatSigned(minor: Long, currency: String = DEFAULT_CURRENCY, locale: Locale = Locale.getDefault()): String {
        val magnitude = formatCompact(abs(minor), currency, locale)
        return when {
            minor > 0 -> "+$magnitude"
            minor < 0 -> "−$magnitude"
            else -> magnitude
        }
    }

    /** What goes in a text field the user then edits: a plain "2575.00", no symbol. */
    fun toEditable(minor: Long, currency: String = DEFAULT_CURRENCY): String {
        val exponent = exponentOf(currency)
        if (exponent == 0) return minor.toString()
        val negative = minor < 0
        val digits = abs(minor).toString().padStart(exponent + 1, '0')
        val whole = digits.dropLast(exponent)
        val fraction = digits.takeLast(exponent)
        return "${if (negative) "-" else ""}$whole.$fraction"
    }

    /**
     * Parses what the user typed. Returns null rather than throwing or guessing — the
     * caller shows a field error instead of quietly saving a wrong amount.
     */
    fun parseOrNull(text: String, currency: String = DEFAULT_CURRENCY): Long? {
        val cleaned = text.trim().replace(" ", "").replace(" ", "")
            .replace("₹", "").replace(",", "")
        if (cleaned.isEmpty()) return null
        val match = Regex("^(-)?(\\d*)(?:\\.(\\d{0,6}))?$").matchEntire(cleaned) ?: return null
        val (sign, whole, fractionRaw) = match.destructured
        if (whole.isEmpty() && fractionRaw.isEmpty()) return null

        val exponent = exponentOf(currency)
        val fraction = fractionRaw.padEnd(exponent, '0')
        val kept = fraction.take(exponent)
        // More precision than the currency has: round half up.
        val roundUp = fraction.length > exponent && fraction[exponent].digitToIntOrNull()?.let { it >= 5 } == true

        val wholeValue = (whole.ifEmpty { "0" }).toLongOrNull() ?: return null
        val unit = Math.pow(10.0, exponent.toDouble()).toLong()
        val value = wholeValue * unit + (kept.ifEmpty { "0" }).toLong() + if (roundUp) 1 else 0
        return if (sign == "-") -value else value
    }

    /** Basis points as a percentage string: 2500 → "25%", 3333 → "33.33%". */
    fun formatBasisPoints(bp: Int): String {
        val whole = bp / 100
        val remainder = bp % 100
        return when {
            remainder == 0 -> "$whole%"
            remainder % 10 == 0 -> "$whole.${remainder / 10}%"
            else -> "$whole.${remainder.toString().padStart(2, '0')}%"
        }
    }

    /**
     * Splits `totalMinor` in the given weight ratio with no minor unit lost or
     * invented, by the same largest-remainder method the server and packages/shared
     * use. The client needs it to preview a split before saving one, and the preview
     * must agree with what the server will compute to the paisa.
     */
    fun apportion(totalMinor: Long, weights: List<Long>, preferIndex: Int = 0): List<Long> {
        if (weights.isEmpty()) return emptyList()
        if (weights.any { it < 0 }) return weights.map { 0L }

        val weightTotal = weights.sum()
        if (weightTotal == 0L) return apportion(totalMinor, weights.map { 1L }, preferIndex)

        val negative = totalMinor < 0
        val magnitude = abs(totalMinor)

        val exact = weights.map { magnitude.toDouble() * it / weightTotal }
        val shares = exact.map { kotlin.math.floor(it).toLong() }.toMutableList()
        var leftover = magnitude - shares.sum()

        val order = exact.indices.sortedWith(
            compareByDescending<Int> { exact[it] - kotlin.math.floor(exact[it]) }
                .thenBy { if (it == preferIndex) 0 else 1 }
                .thenBy { it }
        )

        var cursor = 0
        while (leftover > 0 && order.isNotEmpty()) {
            val index = order[cursor % order.size]
            if (weights[index] > 0 || weights.all { it == 0L }) {
                shares[index] = shares[index] + 1
                leftover -= 1
            }
            cursor += 1
            // Guard against an impossible loop if every weight is zero.
            if (cursor > order.size * (magnitude + 1)) break
        }

        return if (negative) shares.map { -it } else shares
    }
}
